### PR TITLE
add more parameter to DLEstimator/DLClassifier

### PR DESCRIPTION
**Pull request by hhbyyh**
## What changes were proposed in this pull request?

1. add optimMethod, learning rate and learningRateDecay parameter.
2. use SampleToMiniBatch to finish the toBatch operation.
3. change the modifier of SampleToMiniBatch

## How was this patch tested?

extending the existing unit tests

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala [hunk 1]
```diff
@@ -306,7 +306,7 @@ class SampleToBatch[T: ClassTag]
 /**
  * Convert a sequence of [[Sample]] to a sequence of [[MiniBatch]] through function toMiniBatch.
  */
-private[bigdl] class SampleToMiniBatch[T: ClassTag](
+class SampleToMiniBatch[T: ClassTag] private[bigdl](
```
Comment (hhbyyh) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala]: cc @qiuxin2012 for the change.

**Comment by hhbyyh** [comment]
cc @wzhongyuan @jenniew for review.

**Comment by hhbyyh** [comment]
http://172.168.2.101:8080/job/BigDL-PR-Validation/1436/ 
PR validation passed.

**Comment by hhbyyh** [comment]
ping @wzhongyuan for review since it's been 10 days.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala (hunk 1 above)
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala]: just curious, what's the benefit if we put private after the class name ?

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 2]
```diff
@@ -67,7 +68,19 @@ class DLEstimator[@specialized(Float, Double) T: ClassTag](
   def setBatchSize(value: Int): this.type = set(batchSize, value)
 
   /**
-   * number of max Epoch for the training
+   * optimization Method to be used. BigDL supports many optimization methods like Adam,
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: Method => method

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 3]
```diff
@@ -84,14 +119,17 @@ class DLEstimator[@specialized(Float, Double) T: ClassTag](
   protected override def internalFit(
       featureAndLabel: RDD[(Seq[AnyVal], Seq[AnyVal])]): DLModel[T] = {
     val batches = toMiniBatch(featureAndLabel)
-    val dataset = DataSet.rdd(batches)
 
-    val optimizer = Optimizer(model, dataset, criterion)
-      .setOptimMethod(new LBFGS[T]())
+    if(!isDefined(optimMethod)) {
+      set(optimMethod, new LBFGS[T])
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: in Optimizer, SGD is used by default, do we need to keep consistent or any special case here we should use LBFGS ?

Comment (yiheng) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: We should use SGD as default.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala (hunk 1 above)
Comment (hhbyyh) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala]: IMO, It's making the constructor private[bigdl] but not the whole class.

**Comment by hhbyyh** [comment]
Thanks. Updated to address the comments.
PR validation passed: http://172.168.2.101:8080/job/BigDL-PR-Validation/1494/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala (hunk 1 above)
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala]: Got it, thanks

**Comment by wzhongyuan** [comment]
LGTM

**Comment by hhbyyh** [comment]
Thanks for the review.

http://172.168.2.101:8080/job/BigDL-PR-Validation/  passed. Merging it to master
